File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
         in_degrees = [0 for _ in range(numCourses)]
        edges = [[] for _ in range(numCourses)]
        order = []

        for prereq in prerequisites:
            edges[prereq[1]].append(prereq[0])
            in_degrees[prereq[0]] += 1
        
        starts = [course for course in range(numCourses) if in_degrees[course] == 0]

        while starts:
            cur = starts.pop()
            order.append(cur)

            for nxt in edges[cur]:
                in_degrees[nxt] -= 1
                if in_degrees[nxt] == 0:
                    starts.append(nxt)  
        
        return order if len(order) == numCourses else []
        

File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        //kahn's algo
        // find topological sort order
        vector<int> order;
        vector<vector<int>> graph(numCourses); // adjancy list
        vector<int> inDegree(numCourses, 0);
        queue<int> q;

        for (const auto & preReq: prerequisites) {
            int course = preReq[0];
            int preCourse = preReq[1];

            graph[preCourse].push_back(course);
            inDegree[course]++;
        }

        for (int i = 0; i < numCourses; ++i) {
            if (inDegree[i] == 0) {
                q.push(i);
            }
        }

        while (!q.empty()) {
            auto course = q.front();
            q.pop();

            order.push_back(course);

            for (auto nextCourse: graph[course]) {
                inDegree[nextCourse]--;

                if (inDegree[nextCourse] == 0) {
                    q.push(nextCourse);
                }
            }
        }
        return order.size() == numCourses ? order : vector<int>();
    }
};
